proposals/p6333.md

```diff
@@ -0,0 +1,586 @@
+flag names right here, but also we wouldn't expect a proposal for flag name
+changes.
+
+## Details
+
+### Command changes
+
+#### Compile command
+
+The `carbon compile` command is intended to be a straightforward single input,
+single output command. Dependencies will be provided through a combination of:
+
+-   Given a package name to directory mapping, a
+    [filename mapping](#mapping-packaging-directives-to-filenames) based on the
+    library name.
+-   Explicitly provided input files.
+
+As part of supporting a mix of C++ and Carbon files, we will support
+`carbon compile foo.cpp` with results similar to `carbon clang -- -c foo.cpp`.
+
+#### Build command
+
+The `carbon build` command will be the new, simple way to compile, as a
+replacement for `carbon compile`. It will:
+
+-   Load provided files.
+-   For packages with directory mappings, particularly `Core`, figure out the
+    right files to load.
+-   Do something similar to the appropriate series of `carbon compile`
+    invocations.
+    -   A key divergence is that we should avoid re-checking files that would be
+        used across multiple `carbon compile` invocations.
+-   Run the equivalent of `carbon link` over produced inputs.
+
+While the build command will default to providing an executable `program`, we'll
+also want it to be capable of producing `.a` and `.so` files.
+
+#### Link command
+
+The `carbon link` command will change to make the following work:
+
+```sh
+carbon compile foo.carbon -o foo.o
+carbon link foo.o -o program
+```
+
+This requires implicit `compile`-equivalent invocations for `Core` files (not
+just prelude), and includes their object files in output. It should be possible
+to opt out of this, for example so that the Bazel `carbon_binary` rule can use
+`carbon link` while also providing its own `Core` object files. However, it
+should be on-by-default.
+
+### carbon_library and carbon_binary
+
+The Bazel build rules will expose `carbon compile` and `carbon link` behaviors
+in a slightly more Bazel-idiomatic way. For example, given:
+
+```bazel
+carbon_library(
+    name = "lib",
+    srcs = ["a.impl.carbon", "b.impl.carbon", "b.carbon"],
+    apis = ["a.carbon"],
+)
+carbon_binary(
+    name = "bin",
+    srcs = ["main.carbon"],
+    deps = [":lib"],
+)
+```
+
+The way this will approximately work is:
+
+-   `carbon_library` will have an implicit dependency on a set of `Core`
+    libraries (such as a build target `//carbon/lang:core`).
+    -   This will have a network of `carbon_library` rules, some of which may
+        look like `lib`.
+-   For `lib`, we'll delegate to `carbon build`:
+    -   Examine the four files, and determine a compile order.
+    -   Invoke the four `carbon compile` behaviors, producing a `.o` file.
+-   For `bin`, we'll again delegate to `carbon build`:
+    -   Given one file, determine the trivial compile order.
+    -   Invoke `carbon compile` to produce a `.o`
```

Comment:
The phrasing here leaves me unsure whether we're invoking `carbon build`, or whether we're invoking `carbon compile` directly, or some hybrid such as running `carbon build` in a "tell me what you would have done" mode.

Reply:
I think this is a mistake and doesn't reflect how we're likely to want to write this. Rewriting it all. :)